astropy/table/table.py

```diff
@@ -938,7 +938,7 @@ def _set_column_attribute(self, attr, values):
                 if value.strip() == "":
                     value = None
 
-            if value not in (np.ma.masked, None):
+            if value is not None and value is not np.ma.masked:
```

Comment:
Is this stylistic or the source of a subtle bug?

Reply:
Oh, I see the test below. Tricky!

Reply:
It is the source of the bug: `not in list` calls `tuple.__contains__(value)` so the phrasing here is equivalent to `not any(value == t for t in (...))`, while what we want is `not any(value is t for t in (...))` (but for two elements, we might as well expand).

Note that the reason of the failure is somewhat subtle still: `np.ma.masked == unit` drops the mask to check equality of the underlying data. Since `np.ma.masked = MaskedArray(0., mask=True)`, this leads to `np.array(0) == unit`.  Arrays are greedy, so this becomes `np.array(0.) == np.array(unit, dtype=object)`. In the `dtype=object` loop of the `np.equal` ufunc, the individual items get compared, so one ends up with `0. == unit` which failed as `Unit(0.)` creates a unit with zero scale, and in trying to check whether that is the same as `u.one = Unit(1.)`, it is checked whether after decomposing to the same bases (which works), the ratio of scales is unity, so `1./0.` is calculated.

Indeed, the unit fix which forbids zero scale would also solve the table problem on its own, but since `None` and `np.ma.masked` are really meant to be used as singletons, the replacement is still appropriate (and maybe avoids other possibly problematic `value` instances...)

Reply:
Is swapping to `(None, np.ma.masked)` not gonna work? Should we add a comment so future dev won't try to revert this when they attempt to "clean the code"?

Reply:
> Is swapping to `(None, np.ma.masked)` not gonna work?

That would not matter. The comparison with `None` does reliably fail, but then one still has to compare with `np.ma.masked` - I swapped them in my implementation just because `None` seemed more likely to happen than `np.ma.masked` -- though of course almost always `value` is not going to be either of those.

Reply:
Whether a comment would be a good idea: not sure, in the end I felt that the moment it is written as `is`, the code reader will not change it to `not in` because then you do think that uses equality. The problem is really the other way around, that if you want to check whether something is in a list, `not in` seems so logical to use. But happy to add a comment if that is preferred!

Reply:
No, it's okay. With the new tests, I think it is going to be obvious because the tests will fail again. Sorry for the noise... I didn't think it through when I wrote the comment above.